### PR TITLE
Refactor Aggregator and Export data structure

### PR DESCRIPTION
**Pull request by cijothomas**
PR got bigger than I anticipated, but anything touching the Export data structures forces changes in all exporters, and their unit tests. Hoping to finalize the export data types soon, so that Exporter authors can rely on somewhat stable APIs.

Changes:
1. Add enum MetricType to IMetric. This allows exporters to quickly (switch case) determine the output type, and avoids the if..else.ladder of try and cast.

2. Split SumMetricAggregator into two classes one each for Long and Double. (Will come back and see if this copy-paste classes can be avoided with generics). Each class now allocates space for either long or double, instead of having both, but only using one. (Further perf optimizations are possible by determining the aggregators when instrument is created itself, as opposed to doing it in instrument.Record hot path. This will be a follow up PR)

3. Separates the Aggregator (the one having Update/Collect methods) class from the actual Aggregator output. Previously, every Collect() allocated a new instance of Aggregator. This is gotten rid of now, as a single instance of aggregator output is created at ctor(), and re-used for every Collect(). i.e Collect() results in no new allocs from aggregator. (Did this change for only Sum aggregator. Follow up PRs for all other aggregators)

Collect() is still heavy on allocs (Metricitem, Lists from each agg.stores etc.) - I have made it to zero, by borrowing approach similar to one we used inTrace/Log - i.e CircularBuffer+Batch. (I removed these changes from this PR, and it'll be done as an immediate follow up PR, else this would become much much bigger)

4. Minor test stabilization by leveraging PullExporter, instead of speculative Sleep()s

### CODE REVIEW

File: src/OpenTelemetry/Metrics/AggregatorStore.cs [hunk 1]
```diff
@@ -51,9 +51,20 @@ internal IAggregator[] MapToMetrics(string[] seqKey, object[] seqVal)
             var dt = DateTimeOffset.UtcNow;
 
             // TODO: Need to map each instrument to metrics (based on View API)
-            if (this.instrument.GetType().Name.Contains("Counter"))
+            // TODO: move most of this logic out of hotpath, and to MeterProvider's
+            // InstrumentPublished event, which is once per instrument creation.
+
+            if (this.instrument.GetType() == typeof(Counter<long>)
```
Comment (reyang) [inline src/OpenTelemetry/Metrics/AggregatorStore.cs:57]: I hope this can be modeled as enum + union.

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/AggregatorStore.cs:57]: not clear what did you mean here.
(this check will be moved out of the hot paths anyway)

Comment (reyang) [reply src/OpenTelemetry/Metrics/AggregatorStore.cs:57]: e.g.

```
struct Metric
{
    internal AggregationType at; // e.g. Add, TakeTheLastValue, etc.
    public PointType pt; // e.g. int8, int16, uint32, double
    [StructLayout(LayoutKind.Explicit)]
    Point
    {
        [FieldOffset(0)]
        byte uint8;
        [FieldOffset(0)]
        ushort uint16;
        ...
    }
}
```

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/AggregatorStore.cs:57]: got it. Will be part of next rounds :)

**Review by reyang — APPROVED** [review]
LGTM.

File: src/OpenTelemetry/Metrics/MetricAggregators/SummaryMetricAggregator.cs [hunk 2]
```diff
@@ -35,6 +35,7 @@ internal SummaryMetricAggregator(string name, string description, string unit, M
             this.StartTimeExclusive = startTimeExclusive;
             this.Attributes = attributes;
             this.IsMonotonic = isMonotonic;
+            this.MetricType = MetricType.Histogram;
```
Comment (alanwest) [inline src/OpenTelemetry/Metrics/MetricAggregators/SummaryMetricAggregator.cs:38]: `MetricType.Summary`

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs [hunk 3]
```diff
@@ -64,9 +64,10 @@ void ProcessExport(Batch<MetricItem> batch)
                 }
             }
 
+            var processor = new PullMetricProcessor(metricExporter, true);
```
Comment (alanwest) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs:67]: Good call 😄. I wrote these tests first and did not originally think about how the PullMetricProcessor makes things easier.

Comment (cijothomas) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs:67]: I also put TODOs saying provider.Flush() can be leveraged, but then Pull to our rescue!

**Review by alanwest — APPROVED** [review]
LGTM

**Pull request merged**